segmentation_and_clustering.py

```diff
@@ -0,0 +1,105 @@
+    """
+    Segment point cloud using region growing based on normals similarity and spatial proximity.
+    :param point_cloud: Nx3 numpy array of points.
+    :param normals: Nx3 numpy array of corresponding normals.
+    :param angle_threshold: Maximum angle (in degrees) between normals to consider them similar.
+    :param distance_threshold: Maximum spatial distance for points to be considered in the same region.
+    :return: List of clusters, each is a list of indices into the point cloud.
+    """
+    point_count = point_cloud.shape[0]
+    visited = np.zeros(point_count, dtype=bool)
+    clusters = []
+    tree = KDTree(point_cloud)
+
+    # Convert angle threshold from degrees to cosine of the angle for comparison
+    cos_angle_threshold = np.cos(np.radians(angle_threshold))
+
+    for i in range(point_count):
+        if not visited[i]:
+            visited[i] = True
+            cluster = [i]
+            # Use a list as a queue for the region growing
+            queue = [i]
+            while queue:
+                current_point = queue.pop(0)
+                current_normal = normals[current_point]
+
+                # Query for nearby points
+                indices = tree.query_ball_point(
+                    point_cloud[current_point], distance_threshold
+                )
+                for idx in indices:
+                    if (
+                        not visited[idx]
+                        and np.dot(normals[idx], current_normal) >= cos_angle_threshold
+                    ):
+                        visited[idx] = True
+                        queue.append(idx)
+                        cluster.append(idx)
+            clusters.append(cluster)
+
+    return clusters
+
+
+def simulate_parameters(point_cloud, normals, angle_thresholds, distance_thresholds):
+    """
+    Simulate different parameter combinations and plot the resulting clusters.
+    :param point_cloud: Nx3 numpy array of points.
+    :param angle_thresholds: List of angle threshold values to test.
+    :param distance_thresholds: List of distance threshold values to test.
+    """
+    clusters_list = []
+    angle_list = []
+    distance_list = []
+    for angle_threshold in angle_thresholds:
+        for distance_threshold in distance_thresholds:
+            print(
+                f"Simulating with angle threshold={angle_threshold}, distance threshold={distance_threshold}"
+            )
+            clusters = region_growing(
+                point_cloud, normals, angle_threshold, distance_threshold
+            )
+            # plot_clusters_plotly(point_cloud, clusters)
+            print(f"Number of clusters found: {len(clusters)}")
+            clusters_list.append(len(clusters))
+            angle_list.append(angle_threshold)
+            distance_list.append(distance_threshold)
+            # print(clusters)
+
+    return np.stack(
+        (np.array(angle_list), np.array(distance_list), np.array(clusters_list)), axis=1
```

Comment:
Check if np.array is necessary or not to construct 3 dimensional array with np.stack.